k9mail-library/src/main/java/com/fsck/k9/mail/store/imap/selectedstate/command/SelectedStateCommand.java

```diff
@@ -0,0 +1,296 @@
+    calculating the length of the command)
+     */
+    private static final int LENGTH_LIMIT_WITHOUT_CONDSTORE = 980;
+    private static final int LENGTH_LIMIT_WITH_CONDSTORE = 8172;
+
+    ImapConnection connection;
+    ImapFolder folder;
+
+    Set<Long> idSet;
+    List<Range> idRanges;
+
+    SelectedStateCommand(ImapConnection connection, ImapFolder folder) {
+        this.connection = connection;
+        this.folder = folder;
+    }
+
+    abstract String createCommandString();
+
+    abstract Builder newBuilder();
+
+    private List<SelectedStateCommand> splitCommand(int lengthLimit) {
+
+        List<SelectedStateCommand> commands = new ArrayList<>();
+
+        if (idSet != null || idRanges != null) {
+
+            while ((idSet != null && !idSet.isEmpty()) || (idRanges != null && !idRanges.isEmpty())) {
+
+                Builder builder = this.newBuilder()
+                        .idSet(null)
+                        .idRanges(null);
+
+                int length = builder.build().createCommandString().length();
+                while (length < lengthLimit) {
+
+                    if (idSet != null && !idSet.isEmpty()) {
+
+                        Long first = idSet.iterator().next();
+                        length += (String.valueOf(first).length() + 1);
+                        if (length < lengthLimit) {
+                            builder.addId(first);
+                            idSet.remove(first);
+                        } else {
+                            break;
+                        }
+
+                    } else if (idRanges != null && !idRanges.isEmpty()) {
+
+                        Range first = idRanges.iterator().next();
+                        length += (first.toString().length() + 1);
+                        if (length < lengthLimit) {
+                            builder.addIdRange(first.getStart(), first.getEnd());
+                            idRanges.remove(first);
+                        } else {
+                            break;
+                        }
+                    } else {
+                        break;
+                    }
+                }
+                commands.add(builder.build());
+            }
+
+        } else {
+            //This should never happen
+            commands = Collections.singletonList(this);
+        }
+
+        return commands;
+
+    }
+
+    void addIds(StringBuilder builder) {
+
+        if (idSet != null || idRanges != null) {
+
+            optimizeGroupings();
+
+            if (idSet != null) {
+                builder.append(ImapUtility.join(",", idSet));
+            }
+            if (idRanges != null) {
+                if (idSet != null) {
+                    builder.append(",");
+                }
+                builder.append(ImapUtility.join(",", idRanges));
+            }
+            builder.append(" ");
+        }
+    }
+
+    public SelectedStateResponse execute() throws MessagingException {
+        return null;
```

Comment:
It's meant to be a method that does not necessarily need to overridden. I saw this pattern somewhere in the k-9 code. The search and copy commands return parsed responses and so override this. The fetch command is never executed as it directly uses the connection object. The store method needs to be executed but the response is not used.